### PR TITLE
[BRE-834] Prerequisite Stages

### PR DESCRIPTION
**Pull request by ragurney**
Allows users to select prerequisite stages.

screen shot 2018-07-24 at 4 20 09 pm
screen shot 2018-07-24 at 3 17 45 pm
screen shot 2018-07-24 at 4 32 07 pm
screen shot 2018-07-24 at 4 31 58 pm

/cc @zendesk/samson @zendesk/bre

### References
 - Jira link: [BRE-834](https://zendesk.atlassian.net/browse/BRE-834)

### CODE REVIEW

File: app/assets/javascripts/ref_status_typeahead.js [hunk 1]
```diff
@@ -36,8 +37,17 @@ function refStatusTypeahead(options){
     });
   }
 
-  function show_status_problems(status_list) {
+  function show_status_problems(status_list, type) {
     $ref_status_container.removeClass("hidden");
+
+    if (type === 'danger') {
+      $ref_status_container.addClass('alert-danger');
```
Comment (grosser) [inline app/assets/javascripts/ref_status_typeahead.js]: use toggleClass ?

File: app/assets/javascripts/ref_status_typeahead.js [hunk 2]
```diff
@@ -70,6 +82,10 @@ function refStatusTypeahead(options){
             $tag_form_group.addClass("has-error");
             show_status_problems(response.status_list);
             break;
+          case "fatal":
+            $submit_button.prop("disabled", true);
+            show_status_problems(response.status_list, 'danger');
```
Comment (grosser) [inline app/assets/javascripts/ref_status_typeahead.js]: would a boolean make more sense ?

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -289,9 +289,10 @@ def paginate(pagy)
...
 
-  def list_with_show_more(items, display_limit, show_more_tag, ul_options = {}, &block)
-    li_tags = items.first(display_limit).map { |i| capture(i, &block) }
-    li_tags << show_more_tag if items.size > display_limit
+  def unordered_list(items, display_limit: nil, show_more_tag: nil, ul_options: {}, li_options: {}, &block)
+    shown_items = items.first(display_limit || items.size)
+    li_tags = shown_items.map { |item| content_tag(:li, '', li_options) { capture(item, &block) } }
+    li_tags << show_more_tag if display_limit && items.size > display_limit
 
     content_tag(:ul, li_tags.join.html_safe, ul_options)
```
Comment (grosser) [inline app/helpers/application_helper.rb]: should use a html-safe-join ... safe_join or append to `"".html_safe`

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -289,9 +289,10 @@ def paginate(pagy)
     result
   end
 
-  def list_with_show_more(items, display_limit, show_more_tag, ul_options = {}, &block)
-    li_tags = items.first(display_limit).map { |i| capture(i, &block) }
-    li_tags << show_more_tag if items.size > display_limit
+  def unordered_list(items, display_limit: nil, show_more_tag: nil, ul_options: {}, li_options: {}, &block)
+    shown_items = items.first(display_limit || items.size)
+    li_tags = shown_items.map { |item| content_tag(:li, '', li_options) { capture(item, &block) } }
```
Comment (grosser) [inline app/helpers/application_helper.rb]: don't need the second argument ?

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -289,9 +289,10 @@ def paginate(pagy)
     result
   end
 
-  def list_with_show_more(items, display_limit, show_more_tag, ul_options = {}, &block)
-    li_tags = items.first(display_limit).map { |i| capture(i, &block) }
-    li_tags << show_more_tag if items.size > display_limit
+  def unordered_list(items, display_limit: nil, show_more_tag: nil, ul_options: {}, li_options: {}, &block)
+    shown_items = items.first(display_limit || items.size)
```
Comment (grosser) [inline app/helpers/application_helper.rb]: could be simpler to do `items = items.first(display_limit) if display_limit `

File: app/helpers/application_helper.rb [hunk 6]
```diff
@@ -330,4 +331,23 @@ def deployed_or_running_list(stages, reference)
...
+      result << content_tag(:p, help_text, class: 'col-lg-offset-2')
+      result << content_tag(:div, class: 'col-lg-4 col-lg-offset-2') do
+        check_boxes = ''.html_safe
+        check_boxes << collection_check_boxes(object, method, collection, :id, :name) do |b|
+          box = ''.html_safe
+          box << b.check_box + ' '
+          box << b.label
+          box << tag(:br)
+        end
+        check_boxes
```
Comment (grosser) [inline app/helpers/application_helper.rb]: don't need this line

File: db/schema.rb [hunk 7]
```diff
@@ -527,8 +528,8 @@
     t.boolean "builds_in_environment", default: false, null: false
     t.boolean "block_on_gcr_vulnerabilities", default: false, null: false
     t.boolean "notify_assertible", default: false, null: false
-    t.string "notify_email_address"
-    t.float "average_deploy_time"
+    t.datetime "average_deploy_time"
+    t.string "prerequisite_stage_ids"
```
Comment (grosser) [inline db/schema.rb]: should not reorder columns

File: plugins/prerequisite_stages/app/decorators/stage_decorator.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    def prerequisite_stages
+      prerequisite_stage_ids.any? ? Stage.where(id: prerequisite_stage_ids) : []
+    end
+
+    def unmet_prerequisite_stages(reference)
+      commit = project.repository.commit_from_ref(reference)
+      return [] unless commit && prereq_stages = prerequisite_stages.presence
+
+      prereq_stages_with_deployed_ref = prereq_stages.joins(deploys: :job).where(
+      jobs: {status: 'succeeded', commit: commit}
```
Comment (grosser) [inline plugins/prerequisite_stages/app/decorators/stage_decorator.rb]: indent

File: plugins/prerequisite_stages/app/decorators/stage_decorator.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end)
+
+  serialize :prerequisite_stage_ids, Array
+
+  validate :valid_prerequisites?, if: :prerequisite_stage_ids_changed?
+
+  private
+
+  def valid_prerequisites?
```
Comment (grosser) [inline plugins/prerequisite_stages/app/decorators/stage_decorator.rb]: prefer calling these `validate_` for consistency

File: plugins/prerequisite_stages/app/views/samson_prerequisite_stages/_stage_show.html.erb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<% if prerequisite_stages = @stage.prerequisite_stages.presence %>
```
Comment (grosser) [inline plugins/prerequisite_stages/app/views/samson_prerequisite_stages/_stage_show.html.erb]: I think this will do 2 queries since it's .empty + .to_a

Comment (ragurney) [reply plugins/prerequisite_stages/app/views/samson_prerequisite_stages/_stage_show.html.erb]: Tested it out in rails console and it looks like it just does one select

**Comment by grosser** [comment]
but using prerequisite_stages after that will do a full load ... is that
never called ?

**Pull request merged**